added authorize endpoint

sbueringer opened the pull request:
https://github.com/Azure/kubernetes-policy-controller/issues/22

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/kubernetes-policy-controller?pullRequest=28) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/kubernetes-policy-controller?pullRequest=28) before we can accept your contribution.<br/><br/>:x: sbueringer [sign now](https://cla.opensource.microsoft.com/Azure/kubernetes-policy-controller?pullRequest=28)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/kubernetes-policy-controller?pullRequest=28) it.</sub>

rite2nikhil commented:
Thank you for adding this support.  Can you also please add more docs and demo and deployment files. Thanks, then we should be able to merge, also some unit tests will be great

sbueringer commented:
Yep of course. I'll try to get it done this week.

sbueringer commented:
So I've updated my PR with documentation, examples and unit tests. I had to rewrite some of your documentation because a whole additional scenario is now possible. I hope that's ok for you.

So just to give a quick overview what's in this PR:
* Additional documentation for `authorization` scenario with corresponding examples
  * Unfortunately I'm not able to update the visio image with the new authorization scenario as I've currently no Visio.
* I've added the following flags:
  * opa-ca-file
  * opa-auth-token-file
  * log-level
* I've added an additional `authorize` endpoint, which can be configured as authorization module for the Kubernetes APIServer. The whole SubjectAccessReview is propagated to OPA and any deny leads to a deny to Kubernetes
* I've changed the propagation of the AdmissionReview to OPA. It now also propagates the whole AdmissionReview. As far as I know (and I've tested) this shouldn't break the audit use-case because I've updated `matches` as well. It now matches based on the synced resource from `kube-mgmt` and on the data from the `/admit` call.

P.S.: Unfortunately I'm only available until Friday (and then on vacation until January). So it would be nice if we could get this merged until then, but if the PR doesn't make sense or it isn't ready until then it's also no problem. I'll just stay on my fork until I'm back.

rite2nikhil commented on policy/kubernetes/matches.rego:
```diff
@@ -1,6 +1,10 @@
 import data.kubernetes
 
+matches[[kind, namespace, name, resource]] {
+	resource := kubernetes[kind][namespace][name].object
```
Why is this needed ?

sbueringer replied:
I changed the call from the /admit endpoint to the OpenPolicyAgent to add an additional layer. The original resource is send as Object and things UserInfo or OldObject are send with it. So a call to the OPA looks like the following now:
````
data.admission.deny[
  {
    "id": id,
    "resource": {
      "kind": "namespaces",
      "name": "CCbwwpbqch"
    },
    "resolution": resolution
  }
] with data["kubernetes"]["namespaces"]["CCbwwpbqch"] as {
  "uid": "1f5ea330-ec88-11e8-8184-fa163ec0ee41",
  "kind": {
    "group": "",
    "version": "v1",
    "kind": "Namespace"
  },
  "resource": {
    "group": "",
    "version": "v1",
    "resource": "namespaces"
  },
  "operation": "CREATE",
  "userInfo": {
    "username": "cluster-admin",
    "groups": [
      "system:masters",
      "system:authenticated"
    ]
  },
  "object": {
    "metadata": {
      "name": "test",
      "uid": "1f5e9e79-ec88-11e8-8184-fa163ec0ee41",
      "creationTimestamp": "2018-11-20T05:50:02Z"
    },
    "spec": {
      "finalizers": [
        "kubernetes"
      ]
    },
    "status": {
      "phase": "Active"
    }
  }
},
````
This has the advantage that Validating/MutatingWebhooks can also decide based on UserInfo or the oldObject (if that makes sense). Unfortunately the normal machtes does not like before. So I've added another matches: one works now for the /audit scenario and one for /admit and the rules defined in .rego files work independent of its an /admit or /audit use case and do not have to be changed.

rite2nikhil replied:
Makes perfect sense thanks was planing to do that :). Why do you still have the old matches in the package ?

sbueringer replied:
Because I thought kube-mgmt still syncs the Kubernetes resources with the "old" format, so that the rules work for both. But actually I'm not sure ;)

rite2nikhil commented on policy/authorization/psp.rego:
```diff
@@ -0,0 +1,15 @@
+package authorization
+
+deny[{
+	"id": "podsecuritypolicies-kube-system",
```
How will this policy be 'audit' ?

sbueringer replied:
I don't think the 'audit' use case makes sense for the 'authorization' scenario. The usual scenario is that you check with 'audit' if existing resources match to the policies. The authorization calls are only partly create/update on resources and the majority are like get/use/port-forward/exec/logs/... requests and are not leaving any persistent resources in Kubernetes. For example there is no way to write on rule which applies at the same time to:
* The request from Kubernetes to use a specific PodSecurityPolicy (from the kube-controller-manager)
* The then created Pod which now has this PodSecurityPolicy

The /authorization call matches on SubjectAccessReviews from the APIServer, but these SubjectAccessReviews are afaik not persistently stored in Kubernetes and are just transient resources.

rite2nikhil replied:
The scenario is as follows 
An `admin` persona user is adding a new policy to the system and wants to `audit` the affect of the policy before making a change. So in theory he can apply policy on current state of the cluster (based on past requests)

sbueringer replied:
I think that absolutly makes sense for Validating/MutatingWebhooks. I think for the Authorization Use Case that won't work because they match on the SubjectAccessReview and not on the Resources itself (like they to in the Validating/MutatingWebhooks)

rite2nikhil commented on policy/authorization/webhooks.rego:
```diff
@@ -0,0 +1,16 @@
+	"id": "webhookconfigurations-dhc-prefix",
+	"resolution": "You're not allowed to create/update/delete ValidatingWebhookConfigurations & MutatingWebhookConfigurations starting with 'system'",
+}] {
+	input.kind = "SubjectAccessReview"
```
Why not make `SubjectAccessReview` data model follow the same model of 'admit` policies instead of using `input` ?

sbueringer replied:
I guess that;s possible. The only problem is that not every SubjectAccessReview directly matches to the existing pattern (NonResourceAttributes). But as discussed a Mapping is possible and I'll update the PR.

sbueringer commented:
I refactored what we discussed and updated the documentation & examples. Could you please take another look? I also migrated all our internal policies to the new format and I'll verify them in parallel.

sbueringer commented:
I'm not sure how it did work for you but when I tried to send data.authorization.deny Queryies with and without namespace attribute, I got compile errors from OPA, e.g.:
````
data.authorization.deny[{
			"id": id,
			"resource": {"kind": "%s", "namespace": "%s", "name": "%s"},
			"resolution": resolution,}]`,
````
vs.
````
data.authorization.deny[{
			"id": id,
			"resource": {"kind": "%s", "name": "%s"},
			"resolution": resolution,}]`,
````
I now switched to all queries sending a namespace. For non-Resource requests the namespace is just set to an empty string.

sbueringer commented:
@rite2nikhil You can merge the PR if you want. If I'll find any other issues I open further PRs.

yue9944882 commented:
is the webhook authz module reverted? i hit nothing when searching `SubjectAccessReview` in the repo..

maxsmythe commented:
Correct. Authz webhooks cannot rely on the API Server, which the KRM-based model we moved to has.

It is possible to create a standalone server that uses the constraint framework to execute Authz webhooks, but that would require someone to lead that effort.

The old code that supported authz lives here:

https://github.com/open-policy-agent/gatekeeper/tree/v2.0.1

though note that it is unsupported and the Rego works differently than the current model.
